recordCollection.go

```diff
@@ -0,0 +1,196 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/cloudflare/cloudflare-go"
+	"github.com/miekg/dns"
+)
+
+type (
+	recordCollection []cloudflare.DNSRecord
+)
+
+// Find will serach for needle in a recordCollection.
+func (c recordCollection) Find(needle cloudflare.DNSRecord) (int, *cloudflare.DNSRecord) {
+	for i, r := range c {
+		if match(r, needle) {
+			return i, &r
+		}
+	}
+
+	return -1, nil
+}
+
+// Diff will find the differences between two recordCollections.
+func (c recordCollection) Diff(remote recordCollection) (recordCollection, recordCollection) {
+	localOnly := recordCollection{}
+	remoteOnly := recordCollection{}
+
+	for _, l := range c {
+		n, _ := remote.Find(l)
+
+		if n < 0 {
+			localOnly = append(localOnly, l)
+		}
+	}
+
+	for _, r := range remote {
+		n, _ := c.Find(r)
+
+		if n < 0 {
+			remoteOnly = append(remoteOnly, r)
+		}
+	}
+
+	return localOnly, remoteOnly
+}
+
+func (c recordCollection) Print(w io.Writer) {
+	maxName := 0
+	for _, r := range c {
+		if len(r.Name) > maxName {
+			maxName = len(r.Name)
+		}
+	}
+
+	for _, r := range c {
+		name := r.Name + "." + strings.Repeat(" ", maxName-len(r.Name))
+
+		proxied := ""
+		if r.Proxied {
+			proxied = " ; PROXIED"
+		}
+
+		fmt.Fprintf(w, "%s %d %-8s %s%s\n", name, r.TTL, "IN "+r.Type, r.Content, proxied)
+	}
+}
+
+// parseZone will parse a BIND style zone file and return the zone name and
+// a recordCollection.
+func parseZone(r io.Reader) (string, recordCollection, error) {
+	var zoneName string
+	records := recordCollection{}
+
+	for t := range dns.ParseZone(r, "", "") {
+		if t.Error != nil {
+			return "", recordCollection{}, t.Error
+		}
+
+		// Search for zonename while we're at it.
+		soa, found := t.RR.(*dns.SOA)
+		if found {
+			zoneName = strings.Trim(soa.Header().Name, ".")
+		}
+
+		r := newRecord(t)
+		if r != nil {
+			records = append(records, *r)
+		}
+	}
+
+	if zoneName == "" {
+		return "", recordCollection{}, errors.New("Zone name not found")
+	}
+
+	return zoneName, records, nil
+}
+
+// newRecord will instantiate a new cloudflare-compatible DNS record based on
+// a token from our zone file parser.
+func newRecord(in *dns.Token) *cloudflare.DNSRecord {
+	record := &cloudflare.DNSRecord{
+		Name: strings.Trim(in.Header().Name, "."),
+		TTL:  int(in.Header().Ttl),
```

Comment:
Does this not set the TTL of the record to `1` whenever we want the record to be proxied via the magic TTL values?

Reply:
Indeed it does - that matches Cloudflare's internal TTL values, so it makes sense.